src/webui/www/private/scripts/piecesbar.js

```diff
@@ -0,0 +1,266 @@
+/*
+ * Bittorrent Client using Qt and libtorrent.
+ * Copyright (C) 2022  Jesse Smick <jesse.smick@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+'use strict';
+
+if (window.qBittorrent === undefined) {
+    window.qBittorrent = {};
+}
+
+window.qBittorrent.PiecesBar = (() => {
+    const exports = () => {
+        return {
+            PiecesBar: PiecesBar
+        };
+    };
+
+    const STATUS_DOWNLOADING = 1;
+    const STATUS_DOWNLOADED = 2;
+
+    let piecesBarUniqueId = 0;
+    const PiecesBar = new Class({
+        initialize(pieces, parameters) {
+            const vals = {
+                'id': 'piecesbar_' + (piecesBarUniqueId++),
+                'width': 0,
+                'height': 0,
+                'downloadingColor': 'green',
+                'haveColor': 'blue',
+                'borderSize': 1,
+                'borderColor': '#999'
+            };
+
+            if (parameters && $type(parameters) == 'object')
+                $extend(vals, parameters);
+            vals.height = Math.max(vals.height, 12);
+
+            const obj = new Element('div', {
+                'id': vals.id,
+                'class': 'piecesbarWrapper',
+                'styles': {
+                    'border': vals.borderSize + 'px solid ' + vals.borderColor,
+                    'height': vals.height,
+                }
+            });
+            obj.vals = vals;
+            obj.vals.pieces = $pick(pieces, []);
+
+            obj.vals.canvas = new Element('canvas', {
+                'id': vals.id + '_canvas',
```

Comment:
maybe drop the postfix and just use number for id?

Reply:
Kinda same thing as above, since an id needs to be unique on the whole page this helps ensure that this canvas has a unique id, even unique from it's parent.